src/main/java/org/opensearch/security/privileges/ClusterStateMetadataDependentPrivileges.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+package org.opensearch.security.privileges;
+
+import java.util.concurrent.Future;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import org.opensearch.cluster.metadata.Metadata;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.threadpool.ThreadPool;
+
+/**
+ * Abstract super class for classes which need metadata updates from the cluster state. This class implements
+ * asynchronous updates - that means that any subclass needs to be prepared for not having the most up to date
+ * cluster state.
+ */
+public abstract class ClusterStateMetadataDependentPrivileges {
+
+    private static final Logger log = LogManager.getLogger(ClusterStateMetadataDependentPrivileges.class);
+    private Future<?> updateFuture;
+
+    /**
+     * Updates the stateful index configuration asynchronously with the index metadata from the current cluster state.
+     * As the update process can take some seconds for clusters with many indices, this method "de-bounces" the updates,
+     * i.e., a further update will be only initiated after the previous update has finished. This is okay as this class
+     * can handle the case that it do not have the most recent information. It will fall back to slower methods then.
+     */
+    public synchronized void updateClusterStateMetadataAsync(ClusterService clusterService, ThreadPool threadPool) {
+        long currentMetadataVersion = clusterService.state().metadata().version();
+
+        if (currentMetadataVersion <= getCurrentlyUsedMetadataVersion()) {
+            return;
+        }
+
+        if (this.updateFuture == null || this.updateFuture.isDone()) {
+            this.updateFuture = threadPool.generic().submit(() -> {
+                for (int i = 0;; i++) {
+                    if (i > 5) {
+                        try {
+                            // In case we got many consecutive updates, let's sleep a little to let
+                            // other operations catch up.
+                            Thread.sleep(100);
```

Comment:
is 100 ms enough for the catch-up? Should there be another way to handle this? How often will we run into this?

Reply:
The only purpose of this sleep is not to have a tight loop which can only exit on very narrow exit criteria.

To be honest, I do not know how often one will run into this. During cluster startup I believe, it is likely, but still not very likely. If you add logging to the cluster state, you will see that cluster state changes quite often. In many cases, the `updateClusterStateMetadata` will be even then finished before the next cluster change is coming up. Then, it is only relevant if the `updateClusterStateMetadata` method call is slow - that would be the case on clusters with many indices or complicated role definitions.

Reply:
hmm, so we anticipate bootstrap to slow down a tiny bit for clusters with large number of indices? Are we also saying that security plugin initialization is complete before this new data-structure is pre-computed?

Reply:
Everything is done asynchronously - so, it's not like everything else will wait for this process to finish. 

However, yes, the "deal" of this feature is: Let's do as many computations as possible upfront in order to avoid doing them later. 

So, the startup will at least require more CPU resources. However, I would argue that on clusters with many indices the startup will be slow anyway due to the index recovery - so, we are only talking about a very small percentage.

> Are we also saying that security plugin initialization is complete before this new data-structure is pre-computed?

Yes - the code has the ability to fall back to the slower mechanisms if this data structure is not available.